File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
ListNode* rev(ListNode* head) {
    ListNode* temp = head;
    ListNode* prev = nullptr;
    while(temp) {
        ListNode* back = temp -> next;
        temp -> next = prev;
        prev = temp;
        temp = back;
    }
    return prev;
}

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(!head) return true;
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast && fast -> next) {
            fast = fast -> next -> next;
            slow = slow -> next;
        }
        ListNode* rhead = rev(slow);
        ListNode* temp = head;
        ListNode* rtemp = rhead;
        while(rtemp) {
            if(rtemp -> val != temp -> val) {
                rev(rhead);
                return false;
            }
            rtemp = rtemp -> next;
            temp = temp -> next;
        }
        rev(rhead);
        return true;
     }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null) return true;
        Stack<ListNode> stk = new Stack<>();
        ListNode temp = head;
        while(temp != null) {
            stk.push(temp);
            temp = temp.next;
        }
        temp = head;
        while(!stk.empty()) {
            if(stk.peek().val != temp.val) return false;
            stk.pop();
            temp = temp.next;
        }
        return true;
    }
}